File: src/main/scala/com/harmeetsingh13/scala/consumers/SimpleProducer.scala
package com.harmeetsingh13.scala.consumers

import cakesolutions.kafka.KafkaProducer.Conf
import cakesolutions.kafka.{KafkaProducer, KafkaProducerRecord}
import com.typesafe.config.{Config, ConfigFactory}
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.serialization.StringSerializer

import scala.util.{Failure, Success}

/**
  * Created by harmeet on 10/2/17.
  */

class SimpleProducer(config: Config) {

  private val producer = KafkaProducer(
    Conf(config,
      keySerializer = new StringSerializer,
      valueSerializer = new StringSerializer)
  )

  def fireAndForget(record: ProducerRecord[String, String]) = producer.send(record)

  def asyncSend(record: ProducerRecord[String, String]) = {
    producer.sendWithCallback(record)(record => record match {
      case Success(data) => println(s"Offest: ${data.offset()}, Topic: ${data.topic()}, " +
        s"Partition: ${data.partition()}, Timestamp: ${data.timestamp()} ")
      case Failure(ex) => ex.printStackTrace()
    })
  }

  def close = producer.close()
}

object SimpleProducer {

  def main(args: Array[String]): Unit = {
    val config = ConfigFactory.defaultApplication()
    val simpleProducer = new SimpleProducer(config)

    val topic = config.getString("topic")
    simpleProducer.fireAndForget(KafkaProducerRecord(topic, "Name", "Jimmys"))

    simpleProducer.asyncSend(KafkaProducerRecord(topic, "Age", "25"))

    Thread.sleep(1000)
  }
}

File: src/main/java/com/harmeetsingh13/java/producers/SimpleProducer.java
package com.harmeetsingh13.java.producers;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;

import java.util.Properties;

/**
 * Created by harmeet on 10/2/17.
 */
public class SimpleProducer {

    private static Properties kafkaProps = new Properties();
    private static KafkaProducer<String, String> kafkaProducer;

    static {
        kafkaProps.put("bootstrap.servers", "localhost:9092");
        kafkaProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        kafkaProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        kafkaProducer = new KafkaProducer<>(kafkaProps);
    }


    public static void fireAndForget(ProducerRecord<String, String> record) {
        kafkaProducer.send(record);
    }

    public static void asyncSend(ProducerRecord<String, String> record) {
        kafkaProducer.send(record, (recordMetaData, ex) -> {
            System.out.println("Offset: "+ recordMetaData.offset());
            System.out.println("Topic: "+ recordMetaData.topic());
            System.out.println("Partition: "+ recordMetaData.partition());
            System.out.println("Timestamp: "+ recordMetaData.timestamp());
        });
    }

    public static void main(String[] args) throws InterruptedException {
        ProducerRecord<String, String> record1 = new ProducerRecord<>("CustomerCountry",
        "Record 1", "Japan1"
        );

        ProducerRecord<String, String> record2 = new ProducerRecord<>("CustomerCountry",
                "Record 2", "Punjab1"
        );

        fireAndForget(record1);
        asyncSend(record2);

        Thread.sleep(10000);
    }
}
